message RequestEnvelop {
  required int32 unknown1 = 1;
  optional int64 rpc_id = 3;
  repeated Requests requests = 4;
  optional Unknown6 unknown6 = 6;
  optional fixed64 gps_x = 7;
  optional fixed64 gps_y = 8;
  optional fixed64 gps_z = 9;
  optional AuthInfo auth = 10;
  optional int64 unknown12 = 12;

  message Requests {
    required int32 type = 1;
    optional Unknown3 message = 2;
  }

  message Unknown3 {
    required string unknown4 = 1;
  }

  message Unknown6 {
    required int32 unknown1 = 1;
    required Unknown2 unknown2 = 2;

    message Unknown2 {
      required bytes unknown1 = 1;
    }

  }
  message AuthInfo {
    required string provider = 1;
    required JWT token = 2;

    message JWT {
      required string contents = 1;
      required int32 unknown13  = 2;
    }
  }
}

message ResponseEnvelop {
  required int32 unknown1 = 1;
  optional int64 unknown2 = 2;
  optional string api_url = 3;
  optional Unknown6 unknown6 = 6;
  optional Unknown7 unknown7 = 7;
  repeated Payload payload = 100;

  message Unknown6 {
    required int32 unknown1 = 1;
    required Unknown2 unknown2 = 2;

    message Unknown2 {
      required bytes unknown1 = 1;
    }

  }

  message Unknown7 {
    optional bytes unknown71 = 1;
    optional int64 unknown72 = 2;
    optional bytes unknown73 = 3;
  }

  message Payload {
    required int32 unknown1 = 1;
    optional Profile profile = 2;
  }

  message Profile {
    required int64 unknown1 = 1;
    optional string username = 2;
    optional bytes unknown7 = 7; 
    optional Unknown8 unknown8 = 8; 
    optional int32 unknown9 = 9;
    optional int32 unknown10 = 10;
    optional string unknown11 = 11;
    optional string unknown12 = 12;
    optional string unknown13 = 13;
    repeated Currency currency = 14;

    message Unknown8 {
      optional int32 unknown2 = 2;
      optional int32 unknown3 = 3;
      optional int32 unknown9 = 9;
      optional int32 unknown10 = 10;
    }

    message Currency {
      required string type = 1;
      optional int32 amount = 2;
    }
  }
}
